### PR TITLE
extend/ENV/super: add `ENV.O3`

### PR DESCRIPTION
**Pull request by carlocab**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

Having this would have been, or would be, useful on multiple occassions:

- Homebrew/homebrew-core#94724
- Homebrew/homebrew-core#136551
- #15372

I think it's time that we added this back.

### CODE REVIEW

**Comment by Slackadays** [comment]
Why was this ever removed?

**Comment by apainintheneck** [comment]
@Slackadays It looks like it was originally deprecated a few years ago in [this PR](https://github.com/Homebrew/brew/pull/10056/files#diff-14be04a3917a60d17539ef59a9727eb2e9b68501a9fb07b366b9214960b419e6) but it seems like those deprecations might be more related to monkey patching `ENV` or at least the specific reasoning wasn't mentioned there.

**Review by apainintheneck — APPROVED** [review]
Fine with these changes because it doesn't change the default optimization level and there are some examples of formulae that behave differently with different optimization levels. We can apply this on a case-by-case basis.

Would it make sense to add this to the docs somewhere?

**Review by MikeMcQuaid — CHANGES_REQUESTED** [review]
Sorry, realised I had a review here that's been `pending` for a good few days.

File: Library/Homebrew/extend/ENV/super.rb [hunk 1]
```diff
@@ -356,26 +356,15 @@ def refurbish_args
...
-    end
-  end
-
-  sig { params(block: T.nilable(T.proc.void)).void }
-  def O1(&block)
-    if block
-      with_env(HOMEBREW_OPTIMIZATION_LEVEL: "O1", &block)
-    else
-      self["HOMEBREW_OPTIMIZATION_LEVEL"] = "O1"
+  %w[O3 O2 O1 O0 Os].each do |opt|
```
[the remark this answers is not in the document]
Comment (Slackadays) [reply Library/Homebrew/extend/ENV/super.rb:359]: If we don't get O2, what will happen to the `HOMEBREW_OPTIMIZATION_LEVEL` workaround for Clipboard?

File: Library/Homebrew/extend/ENV/super.rb [hunk 2]
```diff
@@ -356,26 +356,15 @@ def refurbish_args
...
+    define_method opt do |&block|
```
[the remark this answers is not in the document]
Comment (apainintheneck) [reply Library/Homebrew/extend/ENV/super.rb:360]: Good point, I remember that being tough when I started with Ruby. Beyond that editors have better support for finding methods when they're defined directly.

File: Library/Homebrew/extend/ENV/super.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/extend/ENV/super.rb:359]: Use O3, O1 or O0.

[the remark this answers is not in the document]
Comment (carlocab) [reply Library/Homebrew/extend/ENV/super.rb]: Done!

File: Library/Homebrew/extend/ENV/super.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (carlocab) [reply Library/Homebrew/extend/ENV/super.rb]: Also done.

File: Library/Homebrew/extend/ENV/super.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (carlocab) [reply Library/Homebrew/extend/ENV/super.rb]: Note, however, that I added `O2` here because that's what was used/needed for Homebrew/homebrew-core#94724.

**Review by MikeMcQuaid — APPROVED** [review]
Looks good, thanks again @carlocab!

**Pull request merged**